### PR TITLE
[spinel] move the `MultiFrameBuffer` implementation to multi_frame_buffer.hpp

### PR DESCRIPTION
**Pull request by zhanglongxia**
The `MultiFrameBuffer` is not only used by `hdlc.hpp`, it is also used by `spi_interface.hpp` and `radio_spinel.hpp`. This commit moves the  `MultiFrameBuffer` implementation from `hdlc.hpp` to `multi_frame_buffer.hpp` as a dependent module.

### CODE REVIEW

**Review by abtink — APPROVED** [review]
LGTM. Thanks 👍

File: src/lib/spinel/multi_frame_buffer.hpp [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+    bool CanWrite(uint16_t aWriteLength) const { return (mRemainingLength >= aWriteLength); }
+
+    /**
+     * Writes a byte into the buffer and updates the write pointer (if space is available).
+     *
+     * @retval OT_ERROR_NONE     Successfully wrote the byte and updated the pointer.
+     * @retval OT_ERROR_NO_BUFS  Insufficient buffer space to write the byte.
+     *
+     */
+    otError WriteByte(uint8_t aByte)
```
Comment (jwhui) [inline src/lib/spinel/multi_frame_buffer.hpp:85]: Document `aByte`.

File: src/lib/spinel/multi_frame_buffer.hpp [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+    }
+
+    uint8_t *mWritePointer;    ///< A pointer to current write position in the buffer.
+    uint16_t mRemainingLength; ///< Number of remaining bytes available to write.
+};
+
+/**
+ * Defines a template frame buffer of a given size for storing a single frame.
+ *
+ * The template parameter `kSize` specifies the size of the buffer.
```
Comment (jwhui) [inline src/lib/spinel/multi_frame_buffer.hpp:121]: Use `@tparam`

File: src/lib/spinel/multi_frame_buffer.hpp [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+    {
+        mWritePointer    = mBuffer;
+        mRemainingLength = sizeof(mBuffer);
+    }
+
+    /**
+     * Indicates whether the buffer is empty or contains a frame.
+     *
+     * @retval TRUE  Buffer is empty
+     * @retval FALSE Buffer contains a frame
```
Comment (jwhui) [inline src/lib/spinel/multi_frame_buffer.hpp:151]:
```suggestion
     * @retval TRUE  Buffer is empty.
     * @retval FALSE Buffer contains a frame.
```

File: src/lib/spinel/multi_frame_buffer.hpp [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+    uint8_t *GetFrame(void) { return mBuffer; }
+
+private:
+    uint8_t mBuffer[kSize];
+};
+
+/**
+ * Defines a template frame buffer of a given size for storing multiple frames.
+ *
+ * The template parameter `kSize` specifies the total size of the buffer.
```
Comment (jwhui) [inline src/lib/spinel/multi_frame_buffer.hpp:179]: Use `@tparam`

**Review by jwhui — APPROVED** [review]

**Pull request merged**